mppsolar/libs/schedule.py

```diff
@@ -0,0 +1,174 @@
+        _loopDuration = config["loop"]
+        _adhocCommandsTopic = config["adhoc_command_topic"]
+
+        _schedules = []
+        for schedule in config["schedules"]:
+            _scheduleType = schedule["type"]
+            if _scheduleType == CommandScheduleType.LOOP:
+                _loopCount = schedule["loopCount"]
+            elif _scheduleType == CommandScheduleType.TIME:
+                _runTime = schedule["runTime"]
+
+            _commands = []
+            for command in schedule["commands"]:
+                _commands.append(cls.parseCommandConfig(command, mqtt_broker, port))
+            
+            if _scheduleType == CommandScheduleType.LOOP:
+                _schedules.append(LoopCommandSchedule(_loopCount, _commands))
+            else:
+                raise ConfigError(f"Undefined schedule type: {_scheduleType}")
+
+        schedule = Schedule(_schedules, _loopDuration, mqtt_broker, port, _adhocCommandsTopic)
+        return schedule
+        
+        
+        
+
+
+
+
+class CommandScheduleType(StrEnum):
+    LOOP = auto()
+    TIME = auto()
+    ONCE = auto()
+
+class InformalScheduledCommandInterface:
+    def is_due(self):
+        raise NotImplementedError
+
+class LoopCommandSchedule(InformalScheduledCommandInterface):
+    def __init__(self, loopCount, commands):
+        self.scheduleType = CommandScheduleType.LOOP
+        self.loopCount = loopCount
+        self.commands = commands
+
+        self.currentLoopCount = loopCount # Set to loopCount so that the first loop will run
+
+    def __str__(self):
+        return f"ScheduleType: {self.scheduleType}, LoopCount: {self.loopCount}, Commands: {self.commands}"
+
+    def is_due(self):
+        if self.currentLoopCount < self.loopCount:
+            self.currentLoopCount += 1
+            return False
+        else:
+            self.currentLoopCount = 1
+            return True
+        
+
+class OneTimeCommandSchedule(InformalScheduledCommandInterface):
+    def __init__(self, command):
+        self.scheduleType = CommandScheduleType.ONCE
+        self.commands = [command]
+        self.has_run = False
+    
+    def __str__(self):
+        return f"ScheduleType: {self.scheduleType}, Commands: {self.commands}"
+    
+    def is_due(self):
+        if self.has_run:
+            log.debug("One time command has already run")
+            return False
+        else:
+            log.debug("One time command is due to run")
+            self.has_run = True
+            return True
+
+
+class Command:
+    def __init__(self, command, commandType, outputs, port):
+        self.command = command
+        self.commandType = commandType
+        self.outputs = outputs
+        self.port = port
+
+    def __str__(self):
+        return f"Command: {self.command}, CommandType: {self.commandType}, Outputs: {self.outputs}"
+    
+    def run(self):
+        log.debug(f"Running command: {self.command}")
+        self.port.connect()
```

Comment:
this is a problem in a couple of porttype, ble doesnt like repeating connects/disconnects, victron shunt and ble continuously send data

Reply:
USB required a new connect with every call. We will need to move the connection state into the device object so the different device types can handle it whichever way is more appropriate.